peer/msg_router.go

```diff
@@ -0,0 +1,270 @@
+package peer
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"golang.org/x/exp/maps"
+)
+
+var (
+	// ErrDuplicateEndpoint is returned when an endpoint is registered with
+	// a name that already exists.
+	ErrDuplicateEndpoint = fmt.Errorf("endpoint already registered")
+
+	// ErrUnableToRouteMsg is returned when a message is unable to be
+	// routed to any endpoints.
+	ErrUnableToRouteMsg = fmt.Errorf("unable to route message")
+)
+
+// EndPointName is the name of a given endpoint. This MUST be unique across all
+// registered endpoints.
+type EndPointName = string
+
+// MsgEndpoint is an interface that represents a message endpoint, or the
+// sub-system that will handle processing an incoming wire message.
+type MsgEndpoint interface {
+	// Name returns the name of this endpoint. This MUST be unique across
+	// all registered endpoints.
+	Name() EndPointName
+
+	// CanRouteTo returns true if the target message can be routed to this
+	// endpoint.
+	CanRouteTo(msg lnwire.Message) bool
+
+	// HandleMessage handles the target message, and returns true if the
+	// message was able to be processed.
+	SendMessage(msg lnwire.Message) bool
```

Comment:
any reason not to combine these two? 

So like `SendMessage(lnwire.Message) bool, bool` 

where the first bool is "success" and the second one is "handled"? 

Just thinking about the case where the underlying impl is dynamic and could return true for `CanRouteTo` but then something changes between when that is called and when `SendMessage` is called

Reply:
I guess it depends on the type of behavior we want for the router: should it be cautious, or optimistic? This also depends on if we want to allow multiple endpoints to handle the same message or not. 


> Just thinking about the case where the underlying impl is dynamic and could return true for CanRouteTo but then something changes between when that is called and when SendMessage is called

IIUC, I think the behavior is the same if we have it be two phase, vs the optimistic single shot:

Two phase:
  * We call `CanRouteTo`, that returns true. 
  * We try to send send the message via `SendMessage` that returns false. 
  * Fail back to caller of `RouteMsg` with an error. 

Single shot:
   * We call `SendMessage`, that calls `CanRouteTo` internally (how else does it know if it can handle the message or not?). That returns false. 
   * Fail back to caller of `RouteMsg` with an error.

Reply:
FWIW I don't have any use case in mind _today_ where a multiple sub-systems would handle a single message. 

However, if we want to _explicitly_ prevent that, then I think we need to make a decision re ordering as:
  * With the way registration works today, you don't know ahead of time if two sub-systems would _potentially_ want the same message. 
  * You can't base it off just the message type as there'll be many co-op close state machines active for each channel undergoing shutdown. 
 
Given the above, we'd either need to just short circuit things (first one that can route, gets it, need to maintain registration order now), or create some sort of unique keys for each `(channel, msg)` pair and base uniqueness off of that. 

Given we don't yet have a scenario in which a multiple sub-systems would accept the same message, I'm inclined to stick with the current design, and update the `godoc` strings to note that there're no guarantees re unique handling.

Reply:
> The subsystem lists all of the messages it handles.

I don't think that's sufficient, as if we have `N` channels doing concurrent co-op closes, we'd have `N` endpoints that all want a `shutdown` message, but one _scoped_ to a particular channel. So in this case, you'd have `(shutdown, chanID)` as the unique key. 

I started to go down this line a bit to improve the routing performance (map look up vs iteration), but it felt very premature, as _dispatching_ likely won't be the bottleneck, and wanted to see how things are used more in the wild before restricting the application space.

Reply:
Ah I see what's going on now. You want the message router to not just route based off of message type but also based off of channel id too. In that case, I think you're right. However, I would love to get to a point where all messages regardless of type, but that are scoped to a channel id have one top level system they run through. This is the `ChannelLifecycle` idea again.